### PR TITLE
JAVA-913: Redesign IndexMetadata API.

### PR DESCRIPTION
**Pull request by adutra**
See CASSANDRA-10216 for details.

### CODE REVIEW

File: driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java [hunk 1]
```diff
@@ -73,167 +95,187 @@
...
+            .hasTarget("text_column")
             .asCqlQuery(createValuesIndex);
-        assertThat(index).isSameAs(column.getIndex("text_column_index"));
-        assertThat(column).hasSingleIndex(index);
-        assertThat(column).isSameAs(index.getColumn("text_column"));
+        assertThat(column.getTable()).hasIndex(index);
+    }
+
+    @Test(groups = "short")
+    public void should_create_metadata_for_simple_index_on_mixed_case_column() {
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java]: This one is failing on <= 2.1 for some reason, though looks like it should work...looking into it.

Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java]: Just needed to selectively create the index based on major version much like should_create_metadata_for_index_on_map_values, i.e.: https://gist.github.com/tolbertam/4be3efec48975fe4b4d4

**Comment by tolbertam** [comment]
Added a version requirement check to one of the tests (feel free to squash that when merging).  This looks good to me :+1:

**Comment by tolbertam** [comment]
Added another test (for testing with multiple indexes on same column - not really necessary but figured would be good to add) and also added 'kind' validation.   PR still looks good to me :+1:

**Comment by tolbertam** [comment]
I noticed via [cassandra-dtests](https://github.com/riptano/cassandra-dtest/blob/540b754108fd0ca74aa0adc088c486d84b71cd1f/materialized_views_test.py#L408-L426) that it was possible to add secondary indexes to materialized views.   I noticed that the driver didn't support that yet, so i added it here in 92dabb3 since it seemed pretty straightforward, feel free to back out/change/etc. if it isn't an appropriate change.

Secondary indexes are semi-busted on C\* trunk right now (added comment to [CASSANDRA-9921](https://issues.apache.org/jira/browse/CASSANDRA-9921)), so the test I added currently fails (unless you put a breakpoint before the execute statement and add the index manually twice), but should pass once fixed in C*.

**Comment by adutra** [comment]
lgtm :+1: thanks for finding out that 2is on mvs are allowed!

**Comment by adutra** [comment]
@tolbertam I just pushed a commit reverting indexes on materialized views following the decision in CASSANDRA-9921. Could you kindly review it again?

**Comment by tolbertam** [comment]
Sounds good, taking a look :)

File: driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java [hunk 2]
```diff
@@ -48,192 +52,304 @@
...
+            .asCqlQuery(createKeysIndex);
+
+        assertThat(getIndex("map_all_values_index"))
+            .hasParent(table)
+            .asCqlQuery(createValuesIndex);
+    }
+
+    @Test(groups = "short", enabled = false)
+    @CassandraVersion(major = 3.0)
+    public void should_create_metadata_for_index_on_materialized_view() {
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java]: It'd probably be fine to just delete this test and if support gets added back we can readd it then

File: driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java [hunk 3]
```diff
@@ -48,192 +52,304 @@
...
+                    ?
+                    ", map_full frozen<map<text, int>>,"
+                        + "set_full frozen<set<text>>,"
+                        + "list_full frozen<list<text>>,"
+                    :
+                    "")
+                + "PRIMARY KEY (id, id2));"
+        );
+
+        if(cassandraVersion.compareTo(VersionNumber.parse("3.0.0-rc1")) >= 0) {
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java]: This is no longer needed if MVs dont support indexes.

Comment (adutra) [inline driver-core/src/test/java/com/datastax/driver/core/IndexMetadataTest.java]: Thanks for both comments, I updated the PR.

**Comment by tolbertam** [comment]
looks good to me :+1:

**Pull request merged**